### PR TITLE
IDs are advertised for [build invocation, workspace, user] to allow cross referencing invocations and linking invocations of the same user and workspace

### PR DESCRIPTION
**Pull request by alkemist**
This change brings two new “features”, both internal at this point.

## Scope IDs 

Within a build invocation, IDs for the invocation, workspace and user are now available. The latter two are persisted in the invocation's root project and Gradle user home dir respectively. The build invocation ID is unique for each invocation.

The build scan plugin will read all IDs and associate them with the scan. This will enable forwards/backwards navigation in time for builds of the same workspace and/or user. It also allows linking related scans, which leads to …

## Reused output origin build ID

When task outputs are reused, the build ID of the producer build is advertised if it is known. This is captured by the build scan plugin. This allows linking from a scan that reused outputs, to the scan of the build that produced those outputs, if a scan exists for it.

This applies to incremental build and build cache.

## Notes

No new public API is introduced in this change. Eventually, the build ID will be exposed to users so that they can record it and use it to find build scans. For example, recording the build ID in the metadata of published JARs and using that to find out about the build that produced the artifact later in time, via the build scan.

## CI

https://builds.gradle.org/project.html?projectId=Gradle_Branches&tab=projectOverview&branch_Gradle_Branches=ld-build-ids

Performance test results: https://builds.gradle.org/viewType.html?buildTypeId=Gradle_Branches_Performance_PerformanceTestCoordinatorLinux&branch_Gradle_Branches_Performance=ld-build-ids&tab=buildTypeStatusDiv

### CODE REVIEW

File: incoming/version.json [hunk 1]
```diff
@@ -1,6 +0,0 @@
-{
-    "commitId" : "6e55986a61d04932f8b34698e5719aa3633fd7c5",
-    "buildNumber" : "482",
-    "buildId" : "980027",
-    "timestamp" : "2016-07-23T07:18:37.643Z"
-}
```
Comment (alkemist) [inline incoming/version.json]: I don't understand why this change is showing here.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/ForkingGradleExecuter.java [hunk 2]
```diff
@@ -179,20 +179,22 @@ public File getWorkingDir() {
...
         }).start();
     }
 
-    protected ForkingGradleHandle createGradleHandle(Action<ExecutionResult> resultAssertion, String encoding, Factory<? extends AbstractExecHandleBuilder> execHandleFactory) {
+    protected ForkingGradleHandle createForkingGradleHandle(Action<ExecutionResult> resultAssertion, String encoding, Factory<? extends AbstractExecHandleBuilder> execHandleFactory) {
         return new ForkingGradleHandle(getStdinPipe(), isUseDaemon(), resultAssertion, encoding, execHandleFactory, getDurationMeasurement());
     }
 
     protected ExecutionResult doRun() {
-        return start().waitForFinish();
```
Comment (alkemist) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/ForkingGradleExecuter.java]: This was the fundamental problem.

start() asserted that there were no afterExecute actions, because they aren't supported for async execution. This meant that `gradleExecuter.afterExecute()` bombed with `ForkingGradleExecuter`.

**Comment by alkemist** [comment]
@lptr @big-guy @wolfs You'll want to at least look at the changes around the origin metadata for cached artifacts. This change adds `buildId`.

**Comment by alkemist** [comment]
@lptr @big-guy @wolfs please go ahead and leave your feedback here. I'll address it in a subsequent change set.

File: subprojects/core/src/main/java/org/gradle/cache/internal/SimpleStateCache.java [hunk 3]
```diff
@@ -59,14 +59,51 @@ public void run() {
...
+            private T result;
+
+            private MaybeUpdater(UpdateAction<T> updateAction) {
+                this.updateAction = updateAction;
+            }
+
+            public void run() {
+                T oldValue = deserialize();
+                result = updateAction.update(oldValue);
+                if (oldValue != result) { // note: intentional pointer comparison
```
Comment (fkorotkov) [inline subprojects/core/src/main/java/org/gradle/cache/internal/SimpleStateCache.java]: Hey @ldaley, why is it intensional?

Comment (ldaley) [reply subprojects/core/src/main/java/org/gradle/cache/internal/SimpleStateCache.java]: I think this is a relic of first trying to jam the “maybe” semantics into the existing `update()` method. We could change this. 

Is it causing you problems?

Comment (fkorotkov) [reply subprojects/core/src/main/java/org/gradle/cache/internal/SimpleStateCache.java]: I've seen that a CI sometimes detects changes to `user-id.txt` when it shouldn't since it's running in the same container each time.

And a brief look at the code lead me to this line. I think it might be because strings will be the same but different references (for example one string is in String Pool and the other was created via `new`). In this case `Gradle` re-write `user-id.txt` and that way changes timestamp of the file which makes CI think it was changed.

Not a big issue at all since it rarely happens but I was wondered why this line is intensional instead something like `Objects.equal(oldValue, result)`

Comment (ldaley) [reply subprojects/core/src/main/java/org/gradle/cache/internal/SimpleStateCache.java]: @fkorotkov I'm going to change this for Gradle 4.6 to use logical equality.
